Improve _GatherDependencies performance by pre-caching path lookup in parallel

csyshing opened the pull request:
### Description of Change(s)

This PR is to improve the performance for `UsdImagingDelegate::SetTime` by pre-caching path dependencies in parallel, in our production shot, the time dropped from 1.2 min to be 10.x second in total.

### Fixes Issue(s)
- https://github.com/PixarAnimationStudios/USD/issues/1813

<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [x ] I have submitted a signed Contributor License Agreement

Commit:
Improve _GatherDependencies performance by pre-caching path lookup in parallel

jilliene commented:
Filed as internal issue #USD-7294

tcauchois commented on pxr/usdImaging/usdImaging/delegate.h:
```diff
@@ -659,7 +659,13 @@ class UsdImagingDelegate : public HdSceneDelegate, public TfWeakBase {
+    typedef TfHashMap<SdfPath, SdfPathVector, SdfPath::Hash> _DependenciesCacheMap;
+    _DependenciesCacheMap _dependenciesCacheMap;
+
+    void _GatherDependenciesCache(SdfPath const &subtree,
```
Can we change this to _CacheDependencies or _PrecacheDependencies?  It makes it more clear what the function is doing.

The pixar code convention is that parameters a function writes out to have to be a pointer, not a reference.  Since the SdfPathVector is owned by the _dependenciesCacheMap, you could change this to SdfPathVector**, or else make the return type "const SdfPathVector&" instead of void...

tcauchois commented on pxr/usdImaging/usdImaging/delegate.h:
```diff
@@ -659,7 +659,13 @@ class UsdImagingDelegate : public HdSceneDelegate, public TfWeakBase {
     _DependencyMap _dependencyInfo;
 
     void _GatherDependencies(SdfPath const& subtree,
-                             SdfPathVector *affectedCachePaths);
```
As with _GatherDependenciesCache, our code convention is to not have data written out to reference parameters.

tcauchois commented on pxr/usdImaging/usdImaging/delegate.h:
```diff
@@ -659,7 +659,13 @@ class UsdImagingDelegate : public HdSceneDelegate, public TfWeakBase {
+                             SdfPathVector& affectedCachePaths);
+
+    typedef TfHashMap<SdfPath, SdfPathVector, SdfPath::Hash> _DependenciesCacheMap;
+    _DependenciesCacheMap _dependenciesCacheMap;
```
I'm worried about this being confused with _dependencyInfo with the current name.  Some other names that could work: _flattenedDependencyCache, _recursiveDependencyCache, etc.

tcauchois commented on pxr/usdImaging/usdImaging/delegate.cpp:
```diff
@@ -994,6 +1006,29 @@ UsdImagingDelegate::ApplyPendingUpdates()
+        // Pre-cache dependencies in parallel
+        WorkDispatcher resyncPathsCacheDispatcher;
+        for (const auto& usdPath: sortedUsdPathsToResync) {
+            auto& affectedCachePaths = _dependenciesCacheMap[usdPath];
```
Building "sortedUsdPathsToResync" will be expensive.  You can delete the loop above by running through "usdPathsToResync", doing the GetPrimPath/GetParentPath/etc, and then replacing this line with:

auto pair = _dependenciesCacheMap.insert(usdPath);
if (pair.second) { continue; /* already inserted */ }
auto& affectedCachePaths = pair->first.second;

tcauchois replied:
Note also that you might want to add caching for "usdPathsToUpdate" as well.  The paths to resync are a result of USD composition/type changes, whereas updates reflect property value changes; both call _GatherDependencies.

tcauchois reviewed (commented):
Thanks so much for the PR!  I've left a few comments for you, on some ideas for performance improvements and then some discussion of names/function signatures.  Would you mind taking a look?

csyshing commented:
Thanks for the review, I will update as soon as I can.

Commit:
Address feedback, fix function declaration to be Pixar's code convention, rename variable for better readability

csyshing commented:
Hi @tcauchois , sorry for the delay, I have addressed the feedback in commit b8a78ea .
This PR was created based on v22.03, I could rebase the change and fast-forward to be based on latest `dev` instead, but let me know if you see any other problem, once it's ready, I could update again.
Please have a look and let me know, cheers!
